src/bootstrap/bootstrap_test.py

```diff
@@ -99,6 +100,52 @@ def test_same_dates(self):
         self.assertFalse(self.build.program_out_of_date(self.rustc_stamp_path, self.key))
 
 
+class ParseArgsInConfigure(unittest.TestCase):
+    """Test if `parse_args` function in `configure.py` works properly"""
+    @patch("configure.err")
+    def test_unknown_args(self, err):
+        # It should be print an error message if the argument doesn't start with '--'
+        configure.parse_args(["enable-full-tools"])
+        err.assert_called_with("Option 'enable-full-tools' is not recognized")
+        err.reset_mock()
+        # It should be print an error message if the argument is not recognized
+        configure.parse_args(["--enable-garbage-collector"])
+        err.assert_called_with("Option '--enable-garbage-collector' is not recognized")
+
+    @patch("configure.err")
+    def test_need_value_args(self, err):
+        """It should print an error message if an argument that requires a value is not given one"""
+        configure.parse_args(["--target"])
+        err.assert_called_with("Option '--target' needs a value (--target=val)")
+
+    @patch("configure.err")
+    def test_option_checking(self, err):
+        # Options should be checked even if `--enable-option-checking` is not passed
+        configure.parse_args(["--target"])
+        err.assert_called_with("Option '--target' needs a value (--target=val)")
+        err.reset_mock()
+        # Options should be checked if `--enable-option-checking` is passed
+        configure.parse_args(["--enable-option-checking", "--target"])
+        err.assert_called_with("Option '--target' needs a value (--target=val)")
+        err.reset_mock()
+        # Options should not be checked if `--disable-option-checking` is passed
+        configure.parse_args(["--disable-option-checking", "--target"])
+        err.assert_not_called()
+
+    @patch("configure.parse_example_config", lambda known_args, _: known_args)
+    def test_known_args(self):
+        # It should call `parse_example_config` with known arguments
+        known_args = configure.parse_args(["--enable-full-tools"])
+        self.assertTrue(known_args["full-tools"][0][1])
+        known_args = configure.parse_args(["--disable-full-tools"])
+        self.assertFalse(known_args["full-tools"][0][1])
+        # It should call `parse_example_config` with known arguments and their values
+        known_args = configure.parse_args(["--target=x86_64-unknown-linux-gnu"])
+        self.assertEqual(known_args["target"][0][1], "x86_64-unknown-linux-gnu")
+        known_args = configure.parse_args(["--target", "x86_64-unknown-linux-gnu"])
+        self.assertEqual(known_args["target"][0][1], "x86_64-unknown-linux-gnu")
```

Comment:
This doesn't seem to test `parse_example_config` at all.

Reply:
It is testing `known_args`. I mocked `parse_example_config` function here because it is parses a TOML file and it is expensive for a unit test. Actually it doesn't need to test this function because it is for `parse_args` function. And it is already covered by some integration tests in the same file.

Reply:
Comments are misleading, that's the issue.

Reply:
But that is the case. `parse_args` is calling `parse_example_config`. I mocked it and that way I can test `known_args`. Comments don't mean they are testing `parse_example_config` but `parse_args` is calling `parse_example_config`.

Reply:
The **unit** test clearly focuses only on the output of `parse_args` with no coverage for `parse_example_config`. If `parse_args` logic changes with keeping the config parsing behavior same, these comments could quickly become unrelated/outdated. In that case, I would either remove or update them.

Reply:
I've updated the comments: https://github.com/rust-lang/rust/pull/132771/commits/de9c37ace9549ea11362112cf69bf0cbf8176b04

BTW, if implementation changes, those tests need to be updated because they actually test _implementation_.

Reply:
> I've updated the comments: https://github.com/rust-lang/rust/pull/132771/commits/de9c37ace9549ea11362112cf69bf0cbf8176b04
> 
> 
> 
> BTW, if implementation changes, those tests need to be updated because they actually test _implementation_.

Only when the desired outcome changes. Otherwise there is no point of having the coverage.